File: sdks/rails/lib/apollo_deploy_signal_sdk_rails/resources/metrics.rb
# frozen_string_literal: true

require_relative "../transport"
require_relative "../errors"

module ApolloDeploySignalSdkRails
  # metrics API domain
  class Metrics
    attr_reader :transport

    def initialize(transport)
      @transport = transport
    end

    # @param id [String]
    # @param query [Hash, nil] query parameters
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [TopicPerformanceResponse]
    def get_topic_performance(id:, query: nil)
      path_params = {
        "id" => id,
      }
      query_params = {}
      query_params["window"] = query[:window] if query && query.key?(:window)
      query_params["projectId"] = query[:project_id] if query && query.key?(:project_id)
      query_params = nil if query_params.empty?

      transport.request(
        method: :get,
        path: "/v1/metrics/topics/${id}",
        path_params: path_params,
        query: query_params,
        timeout_ms: nil,
      )
    end

    # @param id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [EmailPerformanceResponse]
    def get_email_performance(id:)
      path_params = {
        "id" => id,
      }

      transport.request(
        method: :get,
        path: "/v1/metrics/emails/${id}",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

    # @param id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [EmailReadScrollAnalyticsResponse]
    def get_email_engagement(id:)
      path_params = {
        "id" => id,
      }

      transport.request(
        method: :get,
        path: "/v1/metrics/emails/${id}/engagement",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

    # @param query [Hash, nil] query parameters
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [ProjectMetricsSummaryResponse]
    def get_project_metrics_summary(query: nil)
      query_params = {}
      query_params["window"] = query[:window] if query && query.key?(:window)
      query_params["projectId"] = query[:project_id] if query && query.key?(:project_id)
      query_params = nil if query_params.empty?

      transport.request(
        method: :get,
        path: "/v1/metrics/summary",
        query: query_params,
        timeout_ms: nil,
      )
    end

    # @param query [Hash, nil] query parameters
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [ProjectMetricsTimelineResponse]
    def get_project_metrics_timeline(query: nil)
      query_params = {}
      query_params["window"] = query[:window] if query && query.key?(:window)
      query_params["granularity"] = query[:granularity] if query && query.key?(:granularity)
      query_params["format"] = query[:format] if query && query.key?(:format)
      query_params["projectId"] = query[:project_id] if query && query.key?(:project_id)
      query_params = nil if query_params.empty?

      transport.request(
        method: :get,
        path: "/v1/metrics/timeline",
        query: query_params,
        timeout_ms: nil,
      )
    end

    # @param query [Hash, nil] query parameters
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [AdvisorReportResponse]
    def get_metrics_advisor(query: nil)
      query_params = {}
      query_params["projectId"] = query[:project_id] if query && query.key?(:project_id)
      query_params = nil if query_params.empty?

      transport.request(
        method: :get,
        path: "/v1/metrics/advisor",
        query: query_params,
        timeout_ms: nil,
      )
    end

  end
end

File: sdks/python/apollo_deploy_signal_sdk/resources/metrics.py
"""MetricsAPI operations."""

from typing import Any, Dict, Iterator, List, Literal, Optional, Union
from urllib.parse import quote

from pydantic import TypeAdapter
from ..types import AdvisorReportResponse, EmailPerformanceResponse, EmailReadScrollAnalyticsResponse, GetMetricsAdvisorQuery, GetProjectMetricsSummaryQuery, GetProjectMetricsTimelineQuery, GetTopicPerformanceQuery, ProjectMetricsSummaryResponse, ProjectMetricsTimelineResponse, TopicPerformanceResponse
from ..errors import SDKError
from ..transport import RequestOptions, SDKTransport


class MetricsAPI:
    "MetricsAPI API operations"

    def __init__(self, transport: SDKTransport):
        self._transport = transport

    def get_topic_performance(
        self,
        id: str,
        *,
        window: Optional[str] = None,
        project_id: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> TopicPerformanceResponse:
        "getTopicPerformance operation"
        path = "/v1/metrics/topics/${id}"
        path = path.replace("${id}", quote(SDKTransport.wire_value(id), safe=""))
        query: Dict[str, Any] = {}
        if window is not None:
            query["window"] = SDKTransport.wire_query_value(window)
        if project_id is not None:
            query["projectId"] = SDKTransport.wire_query_value(project_id)
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(TopicPerformanceResponse).validate_python(response.json())

    def get_email_performance(
        self,
        id: str,
        *,
        options: Optional[RequestOptions] = None,
    ) -> EmailPerformanceResponse:
        "getEmailPerformance operation"
        path = "/v1/metrics/emails/${id}"
        path = path.replace("${id}", quote(SDKTransport.wire_value(id), safe=""))
        query: Dict[str, Any] = {}
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(EmailPerformanceResponse).validate_python(response.json())

    def get_email_engagement(
        self,
        id: str,
        *,
        options: Optional[RequestOptions] = None,
    ) -> EmailReadScrollAnalyticsResponse:
        "getEmailEngagement operation"
        path = "/v1/metrics/emails/${id}/engagement"
        path = path.replace("${id}", quote(SDKTransport.wire_value(id), safe=""))
        query: Dict[str, Any] = {}
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(EmailReadScrollAnalyticsResponse).validate_python(response.json())

    def get_project_metrics_summary(
        self,
        *,
        window: Optional[str] = None,
        project_id: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> ProjectMetricsSummaryResponse:
        "getProjectMetricsSummary operation"
        path = "/v1/metrics/summary"
        query: Dict[str, Any] = {}
        if window is not None:
            query["window"] = SDKTransport.wire_query_value(window)
        if project_id is not None:
            query["projectId"] = SDKTransport.wire_query_value(project_id)
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(ProjectMetricsSummaryResponse).validate_python(response.json())

    def get_project_metrics_timeline(
        self,
        *,
        window: Optional[str] = None,
        granularity: Optional[str] = None,
        format: Optional[str] = None,
        project_id: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> ProjectMetricsTimelineResponse:
        "getProjectMetricsTimeline operation"
        path = "/v1/metrics/timeline"
        query: Dict[str, Any] = {}
        if window is not None:
            query["window"] = SDKTransport.wire_query_value(window)
        if granularity is not None:
            query["granularity"] = SDKTransport.wire_query_value(granularity)
        if format is not None:
            query["format"] = SDKTransport.wire_query_value(format)
        if project_id is not None:
            query["projectId"] = SDKTransport.wire_query_value(project_id)
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(ProjectMetricsTimelineResponse).validate_python(response.json())

    def get_metrics_advisor(
        self,
        *,
        project_id: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> AdvisorReportResponse:
        "getMetricsAdvisor operation"
        path = "/v1/metrics/advisor"
        query: Dict[str, Any] = {}
        if project_id is not None:
            query["projectId"] = SDKTransport.wire_query_value(project_id)
        headers: Dict[str, str] = {}
        request_options = (options or RequestOptions()).copy()
        if headers:
            request_options.headers.update(headers)
        response = self._transport.request(
            method="GET",
            path=path,
            query=query or None,
            options=request_options,
        )
        if not response.content:
            raise SDKError(
                "Successful response did not contain the expected JSON body",
                response.status_code,
                "empty_response",
                response.headers.get("x-request-id") or response.headers.get("request-id"),
            )
        return TypeAdapter(AdvisorReportResponse).validate_python(response.json())

